### DISCUSSION TITLE
Footer and pages number

### DISCUSSION THREAD

**Discussion by ghost** [dompdf/dompdf#3380] [Q&A]
Hi, 

I am trying for few weeks now to insert a footer with the number of pages. When I started, at least I had the footer even very difficult to display but after so many attempts to get the numero of page, I lost it. I lost everything and no idea what to do.
It looks like $canvas = $dompdf->getCanvas(); doesn't work. I don't understand as I didn't find more recent version of Dompdf.
Is there a library I am missing ?

Thank you

**Comment by bsweeney** [comment]
What does your current logic look like? [`getCanvas()`](https://github.com/dompdf/dompdf/blob/v2.0.4/src/Dompdf.php#L1252-L1260) has been available for a while.

**Comment by ghost** [comment]
It looks like I just forgot to include the canvas library.
But still I have the error : Parse error: syntax error, unexpected token ")" in /.../vendor/dompdf/dompdf/src/PhpEvaluator.php(52) : eval()'d code on line 2.

So I am still stuck

**Reply by bsweeney** [reply]
You're using inlined/embedded script, then? I recommend not doing that from a security perspective.

That error is telling you your PHP syntax is bad. If you provide a sample of your code I can try to help debug.

**Comment by ghost** [comment]
Thank you, 

It's a script used from a WordPress website so I don't know how not to make it embedded.

I am trying to end the code like that : 
```
// Définir la fonction de fermeture
    $footerScript = "
        if ( isset(\$pdf) ) { 
            if (\$PAGE_COUNT > 1) {
                \$font = \$fontMetrics->get_font('Arial, Helvetica, sans-serif', 'normal');
                \$size = 12;
                \$pageText = 'Page ' . \$PAGE_NUM . ' sur ' . \$PAGE_COUNT;
                \$y = 15;
                \$x = 520;
                \$pdf->text(\$x, \$y, \$pageText, \$font, \$size);
            }
        }
    ";

    // Ajouter le script au contenu HTML
    $html .= "<footer>
        <script type='text/php'>{$footerScript}</script>
    </footer>
    </body>
    </html>";

    // Charger le HTML dans Dompdf
    $dompdf->loadHtml($html);

    // Rendre le PDF
    $dompdf->render();

    // Ajouter un pied de page à chaque page
    $canvas = $dompdf->getCanvas();
    $canvas->page_script(function ($pageNumber, Canvas $canvas, Dompdf $dompdf) use ($LogoFooter) {
        $pageText = 'Page ' . $pageNumber . ' sur ' . $canvas->get_page_count();
        $canvas->text(20, 780, "<div style='text-align: center;'><p><a href='https://laksmie.fr/' target='_blank'><img src='data:image/png;base64,{$LogoFooter}' /></a></p></div>");
        $canvas->text(520, 15, $pageText);
    });

    // Sauvegarder le PDF sur le serveur
    file_put_contents($FileName, $dompdf->output());
```

But it still doesn't work

**Reply by bsweeney** [reply]
First, you're already using the canvas to add a page script so there's no need to insert script into your document.

Second, you're mixing up the HTML parsing capabilities of Dompdf with the PDF rendering capabilities of the Canvas. The canvas methods don't accept HTML. So, any text styling, images, links, positioning, etc. have to be done manually. `$canvas->text(20, 780, "<div style='text-align: center;'>...</div>");` doesn't require programmatic insertion so you can just move that HTML into your document. Put it above other content and style `position: fixed;` to have it appears on all pages.

To test out the Canvas method, try starting with something simple like the following:

```
// Charger le HTML dans Dompdf
$dompdf->loadHtml($html);

// Rendre le PDF
$dompdf->render();

// Ajouter un pied de page à chaque page
$canvas = $dompdf->getCanvas();
$canvas->page_script(function ($pageNumber, $pageCount, $canvas, $fontMetrics) {
    $font = $fontMetrics->get_font('helvetica');
    $pageText = 'Page ' . $pageNumber . ' sur ' . $pageCount;
    $canvas->text(520, 15, $pageText, $font, 12);
});

// Sauvegarder le PDF sur le serveur
file_put_contents($FileName, $dompdf->output());
```

If you describe what you're trying to achieve I may be able to provide better guidance.

**Reply by bsweeney** [reply]
fixed this sample

**Comment by ghost** [comment]
Thank you,

I am just trying to print the number of each page, ideally with a logo.
With that : 
```
$footerScript = "
        if ( isset(\$pdf) ) { 
            if (\$PAGE_COUNT > 1) {
                \$font = \$fontMetrics->get_font('Arial, Helvetica, sans-serif', 'normal');
                \$size = 12;
                \$pageText = 'Page ' . \$PAGE_NUM . ' sur ' . \$PAGE_COUNT;
                \$y = 800;
                \$x = 480;
                \$pdf->text(\$x, \$y, \$pageText, \$font, \$size);
            }
        }
    ";

    // Ajouter le script au contenu HTML
    $html .= "<footer>
        <script type='text/php'>{$footerScript}</script>
    </footer>
    </body>
    </html>";

    // Charger le HTML dans Dompdf
    $dompdf->loadHtml($html);
```

I get the last page with the number but no idea where are the others.

**Reply by bsweeney** [reply]
Unless you're defining a page_script, the script will run only on the page where it is defined at the time that page is  being rendered. You appear to be adding it after the last bit of content so it will only appear on the last page. Did you try the page_script method I outlined in my [previous response](https://github.com/dompdf/dompdf/discussions/3380#discussioncomment-8187065)?

Also note, placing script inside a footer won't do much because the script is run then discarded.

**Comment by ghost** [comment]
Yes I tried the previous code with canvas but it didn't work. With a lot of error.

But what does it mean "the script is run then discarded" ? 
I can't have my footer with this page numbers ?

**Reply by bsweeney** [reply]
> Yes I tried the previous code with canvas but it didn't work. With a lot of error.

I mistyped one of the variable references. It's fixed. If you had told me that after you tried it then I could have fixed the sample.

> But what does it mean "the script is run then discarded"?

It means the script is only run one time, not one time per page. HTML is processed separately from PHP, so there is no point mixing the two.

> I can't have my footer with this page numbers?

You can have a footer on every page with page numbers using only HTML and CSS. What you can't have using only HTML/CSS is the page count. To get the total number of pages with current releases you have to use script methods. But when using script you don't have access to the underlying HTML rendering engine, so you have to do all content styling and placement manually.

**Comment by ghost** [comment]
With that : 
```
$canvas = $dompdf->getCanvas();
$pageCount = $canvas->get_page_count();

$font = 'Arial, Helvetica, sans-serif';
$size = 12;

for ($pageNumber = 1; $pageNumber <= $pageCount; $pageNumber++) {
    if ($pageNumber > 1) {
        $pageText = 'Page ' . $pageNumber . ' sur ' . $pageCount;
        $y = 800;
        $x = 480;
        $canvas->text($x, $y, $pageText, $font, $size);
    }
}
```

I get all the numbers on the last page but they seem to be all printed together so it's better but how to print them on each page ?

**Reply by bsweeney** [reply]
At the point you're running this script you're just telling the Canvas to add text to whatever page is currently being rendered. So, yes, running through a loop and calling $canvas->text()` will put all the text on a single page. If you want to run script on each page you should call the page_script method instead. Have you read through the [API documentation](https://github.com/dompdf/dompdf/wiki/Usage#dompdf-api)?

```
$canvas = $dompdf->getCanvas();
$canvas->page_script(function ($pageNumber, $pageCount, $canvas, $fontMetrics) {
    $font = $fontMetrics->get_font('helvetica');
    $size = 12;
    $pageText = 'Page ' . $pageNumber . ' sur ' . $pageCount;
    $y = 800;
    $x = 480;
    $canvas->text($x, $y, $pageText, $font, $size);
});
```

**Comment by ghost** [comment]
Thank you so much. I just tried to add a logo after but once again nothing printed. Idon't understand what I am doing wrong even reading the API 

```
$canvas = $dompdf->getCanvas();
$canvas->page_script(function ($pageNumber, $pageCount, $canvas, $fontMetrics) use ($LogoFooter) {
    $font = $fontMetrics->get_font('helvetica');
    $size = 9;
    $pageText = 'Page ' . $pageNumber . ' sur ' . $pageCount;
    $y = 800;
    $x = 270;
    $canvas->text($x, $y, $pageText, $font, $size);
    
    // Ajouter le logo au pied de page
    $logoX = 450;
    $logoY = 800;
    $logoWidth = 100; 
    $logoHeight = 50; 
    $canvas->image('data:image/jpeg;base64,' . $LogoFooter, $logoX, $logoY, $logoWidth, $logoHeight);
});
```

**Reply by bsweeney** [reply]
The canvas method doesn't support base64-encoded images. You should provide the path to the image file.

**Comment by ghost** [comment]
Thank you from the bottom of my heart.
This footer has been a nightmare for too long.

Do you know also why I can't change the orientation of the page ?
Or should I open another ticket ?

`if ($setPaper == 1) $options->set('defaultPaperOrientation', 'portrait');
    else $options->set('defaultPaperOrientation', 'landscape');`

**Reply by bsweeney** [reply]
It depends on when you're setting that option. You'll want to set the option before instantiating Dompdf since that's the only time the paper size/orientation options are read (#2990 is related). After Dompdf is instantiated, you'll want to use [`$dompdf->setPaper()`](https://github.com/dompdf/dompdf/wiki/Usage#setpaper).

**Reply by ghost** [reply]
I put it before the instantiating. But it doesn't change anything. 
Do you mean that I can use $dompdf->setPaper() after $dompdf = new Dompdf($options) and even change the orientation of the pages in the same file ?
It'd be perfect.

**Reply by bsweeney** [reply]
I'd have to see how you're actually setting the options in the context of Dompdf usage. For example, are you doing something like the following?

```
$options = new Options();
$options->set('defaultPaperOrientation', 'landscape');
$dompdf = new Dompdf($options);
```

Dompdf currently only supports a single size/orientation per document. setPaper doesn't allow you to change the size dynamically, it lets you change the size prior to rendering.

**Reply by ghost** [reply]
Yes That's exactly what I do. And it doesn't work to get a landscape oriented file

**Reply by bsweeney** [reply]
Possible there's anywhere else the paper size might be set?

**Reply by ghost** [reply]
I don't understand the question. Wherever I put the $options->set('defaultPaperOrientation', 'landscape'); It's still in portrait mode.

**Reply by bsweeney** [reply]
I mean is there any other logic that may be setting the page size? Or is the size set through your stylesheet via the `@page` declaration.

**Reply by ghost** [reply]
Yes it was my stylesheet. I'm sorry, I should have thought about it.

Thank yoou for everything
